### PR TITLE
Add fixes for Sandbox extension 

### LINKED ISSUE #2574: Environments Launch splitbutton arrow shows empty flyout for compute systems with no alternative launch options
### Dev Home version

_No response_

### Windows build number

_No response_

### Other software

_No response_

### Steps to reproduce the bug

Install a compute system extension that only supports launch and no other operations.

see
            
            <DataTemplate x:Key="LaunchButton" x:DataType="vm:ComputeSystemViewModel">
                <SplitButton
                        VerticalAlignment="Top"
                        Command="{x:Bind LaunchActionCommand}"
                        x:Uid="ms-resource:///DevHome.Environments/Resources/LaunchButton"
                        Style="{StaticResource CardBodySplitButtonStyle}">
                    <SplitButton.Flyout>
                        <customControls:CardFlyout ItemsViewModels="{x:Bind LaunchOperations}"/>
                    </SplitButton.Flyout>
                </SplitButton>
            </DataTemplate>

### Expected result

The split button arrow should not appear (or at least be disabled?).

### Actual result

The split button arrow can be clicked resulting in an empty flyout.

image

### Included System Information

_No response_

### Included Extensions Information

_No response_

### LINKED ISSUE #3162: Sandbox icon is missing and thumbnail is the wrong size
### Dev Home version

_No response_

### Windows build number

_No response_

### Other software

_No response_

### Steps to reproduce the bug

Have Windows Sandbox enabled
Navigate to environments page

### Expected result

The thumbnail image should be the same size as the others and there should be an icon to the left of "Windows Sandbox"

### Actual result

Looks like the thumbnail is at a different aspect ratio and there's no icon:
image

### Included System Information

_No response_

### Included Extensions Information

_No response_

### PR DESCRIPTION
**Pull request by bbonaby**
## Summary of the pull request
This PR adds updates to the Sandbox extensions. To add the following:

1. Adds a new thumbnail image for the sandboxes environment card. (provided by Design team)
2. Adds the Canary and Stable build constants in the extensions csproj file. This is used to allow the extension to find the correct image icon at run time. See my comment in issue #3162  for more details.
3. Fixes #2574  as it directly affects the sandbox extension. Now we only show the 3 dots for the compute system if it has operations. (this prevents us from showing an empty menu flyout). Also made it so if the split button's menu flyout for the launch button, also has no items, we show a single launch button without the menu flyout.
4. The Sandbox extension can now be enabled/disabled in the extensions library page. So this partially fixes: #3097 . We still need to add the ability to enable the sandbox feature itself from Dev Home just like we do for Hyper-V
5. The Sandbox extension is now disabled when the feature does not exist on the users machine. This happens when we query for extensions that support the compute system provider type in the `ComputeSystemService`.
6. The Sandbox extension now returns the actual error message back to Dev Home to be displayed to the user.

Video showing sandbox interaction in Dev Home: `Note: you can now see initially the extension only has the single launch button. However after it launches we can now see it has a new split button since the Turn off operation is now available`

https://github.com/microsoft/devhome/assets/105318831/a032921e-21fb-4309-b1b1-c9ecc8deb44d

Video showing me now able to enable/disable sandbox extension from within Dev Homes extension library page:

https://github.com/microsoft/devhome/assets/105318831/a9c645ea-6a16-4285-8fcf-9aa1f041caf1

## PR checklist
- [x] Closes #3162
- [x] Closes #2574
- [ ] Tests added/passed
- [ ] Documentation updated

### CODE REVIEW

File: common/Environments/Styles/HorizontalCardStyles.xaml [hunk 1]
```diff
@@ -238,4 +238,14 @@
         <Setter Property="Padding" Value="20,8,15,8" />
     </Style>
 
+    <Style
+        x:Key="CardBodyLaunchButtonStyle"
+
```
Comment (krschau) [inline common/Environments/Styles/HorizontalCardStyles.xaml:243]: Nit: extra line

File: common/Services/ComputeSystemService.cs [hunk 2]
```diff
@@ -102,4 +103,19 @@ public async Task<List<ComputeSystemProviderDetails>> GetComputeSystemProvidersA
 
         return computeSystemProvidersFromAllExtensions;
     }
+
+    /// <summary>
+    /// Gets a boolean indicating whether the extension should be disabled if the Windows optional feature is absent from the machine.
+    /// </summary>
+    /// <param name="extensionClassId">The class Id of the out of proc extension object</param>
+    /// <returns>True when the Windows optional feature is absent. False otherwise.</returns>
+    private bool ShouldDisabledExtensionIfFeatureAbsent(string extensionClassId)
```
Comment (krschau) [inline common/Services/ComputeSystemService.cs:112]:
```suggestion
    private bool ShouldDisableExtensionIfFeatureAbsent(string extensionClassId)
```

Comment (sshilov7) [reply common/Services/ComputeSystemService.cs:112]: I'd propose to move this functionality to ManagementInfrastructureHelper. If we'll have another extensionClassId outside of this file that can be disabled, we'd need to duplicate this code. ManagementInfrastructureHelper can have this ExtensionToFeatureNameMap and answer IsWindowsFeatureAvailable question.

File: common/Services/ComputeSystemService.cs [hunk 3]
```diff
@@ -61,12 +61,13 @@ public async Task<List<ComputeSystemProviderDetails>> GetComputeSystemProvidersA
...
                 }
 
-                // If we're looking at the Hyper-V extension and the feature isn't present on the users machine, disable the extension.
-                // This can happen if the user is not on a SKU that supports Hyper-V.
-                if (extension.ExtensionClassId.Equals(CommonConstants.HyperVExtensionClassId, StringComparison.OrdinalIgnoreCase) &&
-                    ManagementInfrastructureHelper.IsWindowsFeatureAvailable(CommonConstants.HyperVWindowsOptionalFeatureName) == FeatureAvailabilityKind.Absent)
+                // If we're looking at an internal extension that relies upon a Windows optional feature being present on the machine. We need to
+                // check if its on the machine first before allowing it to be added to the list of extensions that are retrieved. If the feature is
+                // absent we disable the extension.
+                if (ShouldDisabledExtensionIfFeatureAbsent(extension.ExtensionUniqueId))
```
Comment (sshilov7) [inline common/Services/ComputeSystemService.cs:67]: How about moving this to ExtensionService to make it more generic? Combined with moving ShouldDisableExtensionIfFeatureAbsent to ManagementInfrastructureHelper (below) other callers would be able to easily use this functionality. And here we would just call 
 DisableExtensionIfFeatureAbsent(extension.ExtensionUniqueId) or
 DisableExtensionIfFeatureAbsent(extension) if "extension" is better for currency here.

**Review by sshilov7 — APPROVED** [review]

File: common/Services/ComputeSystemService.cs (hunk 3 above)
Comment (bbonaby) [reply common/Services/ComputeSystemService.cs]: sounds good, I went ahead and updated it. Thanks!

File: common/Environments/Styles/HorizontalCardStyles.xaml (hunk 1 above)
Comment (bbonaby) [reply common/Environments/Styles/HorizontalCardStyles.xaml]: removed, thanks

File: common/Services/ComputeSystemService.cs (hunk 2 above)
Comment (bbonaby) [reply common/Services/ComputeSystemService.cs]: updated the method name and added Sergeys suggestion

File: common/Helpers/ManagementInfrastructureHelper.cs [hunk 4]
```diff
@@ -64,4 +71,20 @@ private static FeatureAvailabilityKind GetAvailabilityKindFromState(uint state)
...
         }
     }
+
+    /// <summary>
+    /// Gets a boolean indicating whether the extension should be disabled if the Windows optional
+    /// feature is absent from the machine.
+    /// </summary>
+    /// <param name="extensionClassId">The class Id of the out of proc extension object</param>
+    /// <returns>True when the Windows optional feature is absent. False otherwise.</returns>
+    public static bool ShouldDisableExtensionIfFeatureAbsent(string extensionClassId)
```
Comment (dhoehna) [inline common/Helpers/ManagementInfrastructureHelper.cs:81]: NIT:  Change to IsExtensionEnabled.  This changes the method from a negative to a positive.  if(IsExtensionEnabled) is simpler to read than if(!ShouldDisableExtension).  One less negative to think about.

File: extensions/WindowsSandboxExtension/Helpers/Resources.cs [hunk 5]
```diff
@@ -25,7 +26,8 @@ public static string GetResource(string identifier, ILogger? log = null)
                 _resourceLoader = new ResourceLoader(path);
             }
 
-            return _resourceLoader.GetString(identifier);
+            var resourceStr = _resourceLoader.GetString(identifier);
```
Comment (dhoehna) [inline extensions/WindowsSandboxExtension/Helpers/Resources.cs:29]: NIT: rename to resourceStringFormat

File: tools/Environments/DevHome.Environments/Views/LandingPage.xaml [hunk 6]
```diff
@@ -36,16 +36,26 @@
...
-                    </SplitButton.Flyout>
-                </SplitButton>
+                <StackPanel
+                    Visibility="{x:Bind ShouldShowLaunchOperation, Mode=OneWay}">
+                    <SplitButton
+                        Visibility="{x:Bind ShouldShowSplitButton, Mode=OneWay}"
+                        VerticalAlignment="Top"
+                        Command="{x:Bind LaunchActionCommand}"
+                        x:Uid="LaunchButton"
+                        Style="{StaticResource CardBodySplitButtonStyle}">
```
Comment (dhoehna) [inline tools/Environments/DevHome.Environments/Views/LandingPage.xaml:46]: Does this style have color?  If it does it won't respond to theme changes.

**Review by dhoehna — APPROVED** [review]

**Comment by bbonaby** [comment]
Since we're moving the sandbox extension outside of Dev Home, I'll reword this PR as Sergeys suggestions are still good even if Sandbox is removed

**Comment by bbonaby** [comment]
@krschau, I'll make a new PR for this but without the Sandbox related code since there are other things in here that are good to have.

**Comment by krschau** [comment]
@bbonaby Ah sorry, got a little excited closing out PRs. A fresh one is still probably good.

**Pull request closed without merging**